experimental/incremental/task.go

```diff
@@ -518,51 +516,37 @@ func (t *task) run(caller *Task, q *AnyQuery, async bool) (output *result) {
 		defer caller.transferFrom(callee)
 	}
 
-	callee.log("executing", "%T/%v", q.Underlying(), q.Underlying())
-	output.Value, output.Fatal = q.Execute(callee)
-	output.runID = callee.runID
-	callee.log("returning", "%T/%v", q.Underlying(), q.Underlying())
-
-	return output
+	callee.log("executing", "%[1]T/%[1]v", c.query.Underlying())
+	c.value, c.fatal = c.query.Execute(callee)
+	callee.log("returning", "%[1]T/%[1]v", c.query.Underlying())
 }
 
-// waitUntilDone waits for this task to be completed by another goroutine.
-func (t *task) waitUntilDone(caller *Task, async bool) *result {
-	// If this task is being executed synchronously with its caller, we need to
-	// drop our semaphore hold, otherwise we will deadlock: this caller will
-	// be waiting for the leader of this task to complete, but that one
-	// may be waiting on a semaphore hold, which it will not acquire due to
-	// tasks waiting for it to complete holding the semaphore in this function.
-	//
-	// If the task is being executed asynchronously, this function is not
-	// called while the semaphore is being held, which avoids the above
-	// deadlock scenario.
-	if !async {
-		caller.release()
-	}
-
-	select {
-	case <-t.result.Load().done:
-	case <-caller.ctx.Done():
-	}
+type queue[T any] struct {
```

Comment:
You can do better. Pull in this PR https://github.com/bufbuild/protocompile/pull/469.